common.rkt

```diff
@@ -23,50 +25,174 @@
+    (if (var? u)
+        (let ((u-type (var-type-ref u (state-types st))))
+          (if u-type
+              (and (eqv? type? u-type) st)
+              (diseq-simplify (state (state-sub st)
+                                     (state-diseq st)
+                                     (extend-types u type? (state-types st))))))
+        (and (type? u) st))))
+
 
 ;; Reification
-(define (walk* tm sub)
-  (let ((tm (walk tm sub)))
+(struct Ans (term constraint) #:prefab)
+
+(define (walk* tm st)
+  (let* ((sub (state-sub st)) (tm (walk tm sub)))
     (if (pair? tm)
-        `(,(walk* (car tm) sub) .  ,(walk* (cdr tm) sub))
+        `(,(walk* (car tm) st) .  ,(walk* (cdr tm) st))
         tm)))
 (define (reified-index index)
   (string->symbol
     (string-append "_." (number->string index))))
 (define (reify tm st)
   (define index -1)
-  (walk* tm (let loop ((tm tm) (sub (state-sub st)))
-              (define t (walk tm sub))
-              (cond ((pair? t) (loop (cdr t) (loop (car t) sub)))
-                    ((var? t)  (set! index (+ 1 index))
-                               (extend-sub t (reified-index index) sub))
-                    (else      sub)))))
+  (let ((results (let loop ((tm tm) (st st))
+                   (define t (walk tm (state-sub st)))
+                   (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
+                         ((var? t)  (set! index (+ 1 index))
+                                    (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
+                         (else      st)))))
+    (let* ((walked-sub (walk* tm results))
+           (diseq (map (lambda (=/=) (cons (length =/=) =/=)) (state-diseq st)))
+           (diseq (sort diseq (lambda (x y) (< (car x) (car y)))))
+           (diseq (map (lambda (=/=) (cdr =/=)) diseq))
+           (st (diseq-simplify (state (state-sub st) diseq (state-types st))))
+           (diseq (walk* (map pretty-diseq (state-diseq st)) results))
+           (diseq (filter-not contains-fresh? diseq))
+           (diseq (map (lambda (=/=) (sort =/= term<?)) diseq))
+           (diseq (if (null? diseq) '() (list (cons '=/= diseq))))
+           (types (walk* (map pretty-types (state-types st)) results))
+           (types (filter-not contains-fresh? types))
+           (cxs (append types diseq)))
+      (if (null? cxs)
+          walked-sub
+          (Ans walked-sub (sort cxs term<?))))))
+
 (define (reify/initial-var st)
   (reify initial-var st))
+
+(define (term<? u v)
+  (cond
+    ((null? u) (not (null? v)))
+    ((null? v) #f)
+    ((not u) (not v))
+    ((not v) #f)
+    (u v)
+    (v #f)
+    ((number? u) (or (not (number? v)) (< u v)))
+    ((number? v) #f)
+    ((string? u) (or (not (string? v)) (string<? u v)))
+    ((string? v) #f)
+    ((symbol? u) (or (not (symbol? v)) (symbol<? u v)))
+    ((symbol? v) #f)
+    ((pair? u) (or (not (pair? v)) (or (term<? (car u) (car v) (term<? (cdr u) (cdr v))))))
```

Comment:
This should look like:

```
(or (term<? (car u) (car v))
    (and (equal? (car u) (car v))
         (term<? (cdr u) (cdr v))))
```

However, with this approach, a significant amount of redundant checking will occur when two large pairs are equal.  To avoid this redundancy, you can re-implement `term<?` in terms of a 3-result `term-compare` operator instead.

Here is a backwards description of how `term-compare` should behave in terms of `term<?`.  Keep in mind that you can't use `term<?` to implement it, since we're implementing `term-compare` in order to implement `term<?`.  This description is just the most straightforward shorthand I can think of.

```
For (term-compare A B)

when
(term<? A B) return -1
(equal? A B) return  0
(term<? B A) return  1
```